### PR TITLE
[Online demo] engine selector and vidom engine

### PR DESCRIPTION
**Pull request by amel-true**
After #235

### CODE REVIEW

File: common.blocks/demo/demo.js [hunk 1]
```diff
@@ -37,9 +50,11 @@ modules.define('demo', [ 'i-bem__dom', 'pretty', 'functions__debounce' ], functi
         },
         _render: function() {
             try {
-                var api = new BEMHTML({}),
-                    bemhtml = {};
+                var bemhtml = {};
 
+                var api = (this.params.engine === 'vidom')
+                    ? new vidom({})
+                    : new BEMHTML({});
```
Comment (zxqfox) [inline common.blocks/demo/demo.js]:
```js
var Engine = engines[this.params.engine];
new Engine({});
```
?

File: common.blocks/demo/demo.js [hunk 2]
```diff
@@ -5,6 +5,18 @@ modules.define('demo', [ 'i-bem__dom', 'pretty', 'functions__debounce' ], functi
...
+                    this._page = this.findBlockOutside('page');
+                    this._engineSelector = this._page.findBlockInside('engine-selector');
+                    this._$engineSelector = this._engineSelector.domElem;
+
+                    var _this = this;
+                    this.params.engine = this._$engineSelector.find('option:selected').val();
+                    this._$engineSelector.on('change', function() {
+                        _this.params.engine = _this._$engineSelector.find('option:selected').val();
+                        _this._render();
+                    });
```
Comment (nik-kor) [inline common.blocks/demo/demo.js]: .bind(this) ?

File: common.blocks/demo/demo.js (hunk 1 above)
Comment (nik-kor) [inline common.blocks/demo/demo.js]: или 

```js

new (engines[this.params.engine])({})

```

Comment (miripiruni) [inline common.blocks/demo/demo.js]: +1 to @zxqfox

File: common.blocks/demo/demo.js [hunk 3]
```diff
@@ -5,6 +5,18 @@ modules.define('demo', [ 'i-bem__dom', 'pretty', 'functions__debounce' ], functi
             js: {
                 inited: function() {
 
+                    this._page = this.findBlockOutside('page');
+                    this._engineSelector = this._page.findBlockInside('engine-selector');
+                    this._$engineSelector = this._engineSelector.domElem;
+
+                    var _this = this;
```
Comment (miripiruni) [inline common.blocks/demo/demo.js]: `_this` → `demo`

**Comment by miripiruni** [comment]
@amel-true Failed to load resource: the server responded with a status of 404 (Not Found)
`http://localhost:8080/desktop.bundles/index/xindex.browser.vidom.js`

File: desktop.bundles/index/index.bemjson.js [hunk 4]
```diff
@@ -56,10 +57,17 @@ module.exports = {
...
+                        },
+                        {
+                            block: 'engine-selector',
+                            mix: { block: 'header', elem: 'item' },
+                            mods: { state: 'loading' },
+                            versions: [
+                                { name: 'BEMHTML', value: 'bemhtml' },
+                                { name: 'VIDOM', value: 'vidom' },
+                                { name: 'BEMJSON', value: 'bemjson' }
+                            ]
```
Comment (miripiruni) [inline desktop.bundles/index/index.bemjson.js]: 1. BEMHTML
2. BEMJSON
3. VIDOM

`value` unnecessary: `name.toLowerCase()`

File: common.blocks/demo/demo.js [hunk 5]
```diff
@@ -54,9 +69,25 @@ modules.define('demo', [ 'i-bem__dom', 'pretty', 'functions__debounce' ], functi
...
+                case 'vidom':{
+                    compiledText = JSON.stringify(bemhtml.apply(BEMJSON), null, 4);
+                    break;
+                }
+                case 'bemhtml':{
+                    compiledText = pretty(bemhtml.apply(BEMJSON), { max_char: 1000});
+                    break;
+                }
+                case 'bemjson':{
+                    compiledText = JSON.stringify(BEMJSON, null, 4);
```
Comment (zxqfox) [inline common.blocks/demo/demo.js]: Can we make `Engine.render(BEMJSON)` method to incapsulate there logic about rendering for engines?

File: desktop.bundles/index/index.bemjson.js (hunk 4 above)
Comment (amel-true) [inline desktop.bundles/index/index.bemjson.js]: add xindex.browser.vidom.js 04f5d3f1dd2a10e41000df9cb8712d2b2a0ff2b9

Comment (amel-true) [inline desktop.bundles/index/index.bemjson.js]: Yes, it's true 013dab062b3fb15a1dfeca7695f9b974faf00e1b

File: common.blocks/demo/demo.js (hunk 5 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: Something like this 3703c12ef99d050a5f2d9565935e4a506ce923d7?

File: common.blocks/demo/demo.js (hunk 3 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: Removed in 3703c12ef99d050a5f2d9565935e4a506ce923d7

File: common.blocks/demo/demo.js (hunk 2 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: yep 3703c12ef99d050a5f2d9565935e4a506ce923d7

File: common.blocks/demo/demo.js (hunk 1 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: refactoring 3703c12ef99d050a5f2d9565935e4a506ce923d7

File: common.blocks/demo/demo.js (hunk 5 above)
Comment (zxqfox) [inline common.blocks/demo/demo.js]: Yeah, much better. Even better could be a way with classes of Engine interface for each engine. Like:
```js
class BEMXJSTEngine {
  render(bemjson) {
    throw new Error('Doesn\'t .');
  }
  _safe(fn) {
    try {
      return fn();
    } catch(e) {
      return engineName.toUpperCase() + ' error: ' + e.message + '\n' + e.stack;
    }
  }
}
// ...
class BEMHTMLEngine extends BEMXJSTEngine {
  render(bemjson) {
    this.compile(bemjson);
    this.exportApply(bemhtml);
    return this._safe(function() {
      return pretty(bemhtml.apply(BEMJSON), {max_char: 1000});
    });
  }
}
```

**Comment by miripiruni** [comment]
@amel-true can you split this PR to two: 

1) all changes without react engine — and we merge it immediately.
2) changes about react engine — and we do it after #235 

Thank you!

File: common.blocks/demo/demo.js (hunk 5 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: That looks great, but i'm not sure if we can use ES6 syntax?

Comment (zxqfox) [inline common.blocks/demo/demo.js]: Don't worry, we can use ES5 syntax:
```
function BEMXJSTEngine(opts) {} // do we need constructor?
BEMXJSTEngine.prototype.render = function(bemjson) {
  throw new Error("Abstract");
};
// etc.
function BEMHTMLEngine() {}
util.inherits(BEMHTMLEngine, BEMXJSTEngine); // browserify will do the work
BEMHTMLEngine.prototype.render = function(bemjson) {
  // etc.
}
```

Comment (amel-true) [inline common.blocks/demo/demo.js]: Or we can use enb-babel )
aba34dd51dbf036711f205156e74f8a0f92d8b27

File: .enb/make.js [hunk 6]
```diff
@@ -55,13 +57,17 @@ module.exports = function(config) {
...
+                target: '?.es5.js'
+            }],
+            
+            [ techs.prependModules, {
+                source: '?.es5.js',
+                target: '?.browser.js'
             }],
 
             // borschik
-            [techs.borschik, { sourceTarget: '?.js', destTarget: 'x?.js', freeze: true, minify: isProd }],
```
Comment (amel-true) [inline .enb/make.js]: sorry :) da75a4ea420afdc0866a72d9b2384c000404403b

File: common.blocks/demo/demo.js (hunk 5 above)
Comment (amel-true) [inline common.blocks/demo/demo.js]: Yep, i did it  f29434d56022a8d75c8b4115c40a1954727c0aa0

**Comment by amel-true** [comment]
> can you split this PR to two:
> 1) all changes without react engine — and we merge it immediately.
> 2) changes about react engine — and we do it after #235

@miripiruni  i think that this changes doesn't make sense with one engine, but i can disable vidom in engine selector

**Comment by miripiruni** [comment]
@amel-true there is two engines: BEMHTML and BEMTREE

**Comment by amel-true** [comment]
@miripiruni  i disabled vidom in af0b2e3

File: .enb/make.js [hunk 7]
```diff
@@ -55,13 +57,17 @@ module.exports = function(config) {
             [techs.htmlFromBemjson],
 
             // js
-            [techs.browserJs, {
-                target: '?.js',
-                includeYM: true
+            [ techs.babel, {
+                target: '?.es5.js'
```
Comment (miripiruni) [inline .enb/make.js]: For what?

**Comment by miripiruni** [comment]
And what about https://github.com/bem/bem-xjst/pull/238#issuecomment-205265099 comment (4 April).

File: .enb/make.js (hunk 7 above)
Comment (amel-true) [inline .enb/make.js]: For ES6 classer https://github.com/bem/bem-xjst/pull/238/commits/f29434d56022a8d75c8b4115c40a1954727c0aa0#diff-4fe9281de224b8508c7ebda7e35beadfR123

**Comment by amel-true** [comment]
> And what about #238 (comment) comment? (4 April).
@miripiruni https://github.com/bem/bem-xjst/pull/238/commits/6026da6ab18ead57809b4b3d4bc85adb35bf4047

**Comment by miripiruni** [comment]
@amel-true but error still here
screen shot 2016-05-11 at 23 03 16

**Comment by amel-true** [comment]
@miripiruni yep, you need **npm run build** for copying xindex.browser.vidom.js to desktop.bundles/index/. I can add this to readme.md or delete xindex.browser.vidom.js from project.

**Comment by miripiruni** [comment]
> or delete xindex.browser.vidom.js from project.

I prefer to delete all code that relevant vidom until it’s not needed.

**Comment by amel-true** [comment]
@miripiruni done 20a077ba28e52ad78544c7df438d237e1da113c4

**Comment by miripiruni** [comment]
@amel-true к сожалению, я всё ещё не могу это вмёрджить.

1) В рамках задачи не было притаскивать сюда babel.
2) Второй движок BEMTREE (а не BEMJSON).
3) Оно тупо не работает. Ты пробовал написать хоть один шаблон к какому-нибудь BEMJSON-у и попереключать движки?

**Comment by miripiruni** [comment]
ping

**Comment by amel-true** [comment]
Думаю, надо пулл-реквест отменять. Он делался под другую задачу и в итоге перерос во что-то иное. Проще начать с нуля.

**Pull request closed without merging**